documentation/Action-plugin-specification.md

```diff
@@ -0,0 +1,133 @@
+
+## How plugins are loaded.
+
+At startup, the server app scans the `plugins` folder. In each plugin's subfolder, it looks for a `plugin.json` file, from which it reads the metadata. It compares this information with the list of plugins registered in `action_plugin` and updates the database accordingly:
+
+- Adds a new record for newly-found plugins
+- Updates plugin details that have changed
+- Removes records for missing plugins
+
+The `action_plugin` table is the primary record of what actions are available to the system, with the following fields stored:
+
+- `code`
+- `name`
+- `description`
+- `path` (determined by the folder the plugin was found in, plus the specified filename)
+- `function_name`
+- `required_parameters`
+
+<a name="development"></a>
+
+## Development and build process
+
+Plugins are intended to be stand-alone units, in that a user should be able to simply copy a plugin folder to another system and have it work straight away. (Eventually, we envisage that plugins could be imported directly via the front-end UI.) To that end, each plugin is developed as its own package, with its own `package.json` and `node_modules` folder, as well as Typescript and testing configurations.
+
+When developing a plugin, you can run the main app with `yarn dev` as normal, and changes you make to the plugin `.ts` file will be reflected immediately in the dev environment.
+
+However, when development work is complete, the plugin should be built independently by running `yarn build` _in the plugin's root folder_, which will compile the typescript code into `.js` file(s). (There is also a shortcut `yarn build_plugins` to build all plugins from the project root.) When the main project is built (`yarn build`), the plugins are not re-compiled, but simply copied directly to the build folder, so its expected they will each have their own compiled `.js` file already in place.
+
+## Needs consideration
+
+- Should the `required_parameters` specify an expected type so that the Template Builder UI knows what kind of options to present?
+
+- Should there be (optional) **default values** for the `required_parameters`? I'm thinking, for things like "Send notification", the user's email address will be required, but it's always going to be the same query (`{operator: "objectProperties", children: [{value: {object: "user", property: "email"}}]}`) so it seems like it should there by default so the Admin doesn't have to specify an obvious value in the Template Builder.
+
+- Currently, parameters are evaluated immeditely before the Action is saved to the `action_queue`. For immediate actions, this is fine, but for Scheduled Actions, can we imagine a use case where it would be better for them to be evaluated at the time they run (which could be much later, even years)?
```

Comment:
I think it makes all the sense to be evaluated at the time they will run. Otherwise the parameters might have changed. And **I read somewhere** about us storing the queried values so when the action runs it doesn't need to get it from the database. So that enforces even more, the evaluation needs to be right before the action runs.

Any reason not to?

Reply:
Yeah, I guess we just need to try and think of use cases and see which is more likely. 

In terms of code, they're currently evaluated by the Trigger Queue processor and the evaluated results saved into the action queue.

To do it the other way, it makes the scheduled jobs code somewhat more complicated, but since we're not yet sure the best way to handle scheduled actions (I have an prototype, but I'm not sure it's the best way), we should probably have a discussion about this before making new changes.

Reply:
Ok. I agree to leave this one for now as a **To be considered** here